### ISSUE TITLE
Question about densification with split and clone

### ISSUE THREAD

**Issue by leizhenyu-lzy** [graphdeco-inria/gaussian-splatting#1020]
In the source code shown below I can see **a bias is add to the new_xyz when doing split**, but it seems that **when doing clone new gaussian is just using the same properties**, it doesn't look like the image shown in the paper(shown below), and will these 2 gaussians do the same thing(transform in the same way) in the training process and the result will not be the same shown in the picture below?

Thank you for your help and reply.

![image](https://github.com/user-attachments/assets/241c9245-0fb0-4140-b5c1-e1c47a8753e1)

```python
    def densify_and_split(self, grads, grad_threshold, scene_extent, N=2):
        n_init_points = self.get_xyz.shape[0]
        # Extract points that satisfy the gradient condition
        padded_grad = torch.zeros((n_init_points), device="cuda")
        padded_grad[:grads.shape[0]] = grads.squeeze()
        selected_pts_mask = torch.where(padded_grad >= grad_threshold, True, False)
        selected_pts_mask = torch.logical_and(selected_pts_mask,
                                              torch.max(self.get_scaling, dim=1).values > self.percent_dense*scene_extent)

        stds = self.get_scaling[selected_pts_mask].repeat(N,1)
        means =torch.zeros((stds.size(0), 3),device="cuda")
        samples = torch.normal(mean=means, std=stds) 
        rots = build_rotation(self._rotation[selected_pts_mask]).repeat(N,1,1)
        new_xyz = torch.bmm(rots, samples.unsqueeze(-1)).squeeze(-1) + self.get_xyz[selected_pts_mask].repeat(N, 1)  # add bias
        new_scaling = self.scaling_inverse_activation(self.get_scaling[selected_pts_mask].repeat(N,1) / (0.8*N))
        new_rotation = self._rotation[selected_pts_mask].repeat(N,1)
        new_features_dc = self._features_dc[selected_pts_mask].repeat(N,1,1)
        new_features_rest = self._features_rest[selected_pts_mask].repeat(N,1,1)
        new_opacity = self._opacity[selected_pts_mask].repeat(N,1)

        self.densification_postfix(new_xyz, new_features_dc, new_features_rest, new_opacity, new_scaling, new_rotation)

        prune_filter = torch.cat((selected_pts_mask, torch.zeros(N * selected_pts_mask.sum(), device="cuda", dtype=bool)))
        self.prune_points(prune_filter)

    def densify_and_clone(self, grads, grad_threshold, scene_extent):
        # Extract points that satisfy the gradient condition
        selected_pts_mask = torch.where(torch.norm(grads, dim=-1) >= grad_threshold, True, False)
        selected_pts_mask = torch.logical_and(selected_pts_mask,
                                              torch.max(self.get_scaling, dim=1).values <= self.percent_dense*scene_extent)
        
        new_xyz = self._xyz[selected_pts_mask]
        new_features_dc = self._features_dc[selected_pts_mask]
        new_features_rest = self._features_rest[selected_pts_mask]
        new_opacities = self._opacity[selected_pts_mask]
        new_scaling = self._scaling[selected_pts_mask]
        new_rotation = self._rotation[selected_pts_mask]

        self.densification_postfix(new_xyz, new_features_dc, new_features_rest, new_opacities, new_scaling, new_rotation)
```

**Comment by AsherJingkongChen** [comment]
Either those small splats or remaining splats are treated in the same way in the next iteration or later. I think those small splats can have 0.5 std bias.

**Comment by djx99** [comment]
Yes, it seems that when clone, the original gauss is copied exactly `new_xyz = self._xyz[selected_pts_mask]` and an offset is not added. Can anyone explain this issue?


> Either those small splats or remaining splats are treated in the same way in the next iteration or later. I think those small splats can have 0.5 std bias, which would make them different.

Why do they have 0.5 std bias, what does 0.5 std bias mean? Thank you very much.

**Comment by AsherJingkongChen** [comment]
> Yes, it seems that when clone, the original gauss is copied exactly `new_xyz = self._xyz[selected_pts_mask]` and an offset is not added. Can anyone explain this issue?
> 
> > Either those small splats or remaining splats are treated in the same way in the next iteration or later. I think those small splats can have 0.5 std bias, which would make them different.
> 
> Why do they have 0.5 std bias, what does 0.5 std bias mean? Thank you very much.

Just like splitting which has 1.0 deviation.

I think it is OK. You can try it too. The parameters during densification are just heuristics.

**Comment by Xavier-Lin** [comment]
> Yes, it seems that when clone, the original gauss is copied exactly `new_xyz = self._xyz[selected_pts_mask]` and an offset is not added. Can anyone explain this issue?
> 
> > Either those small splats or remaining splats are treated in the same way in the next iteration or later. I think those small splats can have 0.5 std bias, which would make them different.
> 
> Why do they have 0.5 std bias, what does 0.5 std bias mean? Thank you very much.

Although the original Gaussian points were copied exactly during the cloning stage, the corresponding state values of these cloned Gaussians in the optimizer were set to zero (both first-order and second-order moments are zero). During the density control phase, the Gaussian points in the current iteration have not yet undergone gradient-based updates. Therefore, after the cloning operation is completed, when gradient updates are applied, the original Gaussian points will move along the gradient direction, while the cloned Gaussian points will remain unchanged in position. As a result, this formally achieves the cloning of Gaussian points along the gradient direction (though, in reality, the cloned Gaussians are only meant to preserve their original positions).

**Comment by Jwonno** [comment]
> > Yes, it seems that when clone, the original gauss is copied exactly `new_xyz = self._xyz[selected_pts_mask]` and an offset is not added. Can anyone explain this issue?네, 클론을 만들 때는 원본의 가우스 값이 그대로 복사되고, 어떠한 오프셋도 추가되지 않는 것 같습니다. 이 현상에 대해 설명해 주실 수 있나요?
> > > Either those small splats or remaining splats are treated in the same way in the next iteration or later. I think those small splats can have 0.5 std bias, which would make them different.그 작은 얼룩들이나 남아 있는 얼룩들도 다음 반복 과정에서는 동일한 방식으로 처리됩니다. 제 생각에, 그 작은 얼룩들은 0.5의 표준 편차를 가질 수 있으며, 이로 인해 그 얼룩들이 서로 다르게 나타날 수 있습니다.
> > 
> > 
> > Why do they have 0.5 std bias, what does 0.5 std bias mean? Thank you very much.왜 그들은 0.5 표준편차의 편향을 가지고 있을까요? 0.5 표준편차의 편향이란 무엇을 의미할까요? 정말 감사합니다.
> 
> Although the original Gaussian points were copied exactly during the cloning stage, the corresponding state values of these cloned Gaussians in the optimizer were set to zero (both first-order and second-order moments are zero). During the density control phase, the Gaussian points in the current iteration have not yet undergone gradient-based updates. Therefore, after the cloning operation is completed, when gradient updates are applied, the original Gaussian points will move along the gradient direction, while the cloned Gaussian points will remain unchanged in position. As a result, this formally achieves the cloning of Gaussian points along the gradient direction (though, in reality, the cloned Gaussians are only meant to preserve their original positions).비록 클로닝 과정에서 원래의 가우시안 점들이 그대로 복사되긴 하지만, 최적화 과정에서 이러한 복사된 가우시안 점들의 상태 값은 모두 0으로 설정됩니다. 즉, 1차 및 2차 모멘트도 모두 0입니다. 밀도 제어 단계에서는, 현재 반복 과정에서의 가우시안 점들은 아직 그래디언트 기반의 업데이트를 받지 않았습니다. 따라서 클로닝 작업이 완료된 후, 그래디언트 업데이트가 적용되면, 원래의 가우시안 점들은 그래디언트 방향으로 이동하게 되고, 복사된 가우시안 점들은 위치가 그대로 유지됩니다. 이를 통해, 사실상 가우시안 점들이 그래디언트 방향으로 복사된 것처럼 보이지만, 실제로는 복사된 가우시안 점들은 원래의 위치를 그대로 유지하게 됩니다.

Thank you for detail and clear explanation!
Your explanation reaIlly cleared up the part that I was confused about.
So, for making sure I understood correctly, you means that the differences between the original gaussian and the cloned one are the 'momentum' and 'velocity' of the Adam optimizer, right?
